File: sdkwork-models-python/sdkwork_models/query.py
from __future__ import annotations

from .types import ModelCatalog


def list_vendors(catalog: ModelCatalog) -> list[dict]:
    vendors: dict[str, dict] = {}
    for vendor in catalog.vendors:
        vendor_code = vendor["vendorCode"]
        if vendor_code in vendors:
            continue
        vendors[vendor_code] = {
            "vendorCode": vendor_code,
            "displayName": vendor.get("displayName"),
            "legalName": vendor.get("legalName"),
            "vendorType": vendor.get("vendorType"),
            "capabilities": list(vendor.get("capabilities", [])),
            "openSource": vendor.get("openSource", False),
        }
    return list(vendors.values())


def list_vendor_regions(catalog: ModelCatalog) -> list[dict]:
    if catalog.vendor_catalogs:
        return [
            {"vendorCode": item["vendorCode"], "regionCode": item["regionCode"]}
            for item in catalog.vendor_catalogs
        ]
    return [
        {"vendorCode": vendor["vendorCode"], "regionCode": region["regionCode"]}
        for vendor in catalog.vendors
        for region in vendor.get("regions", [])
    ]


def list_models(catalog: ModelCatalog, filter: dict | None = None, **filters: str) -> list[dict]:
    standard_filters = dict(filter or {})
    standard_filters.update(filters)
    result = catalog.models
    if vendor_code := _filter_value(standard_filters, "vendorCode", "vendor_code"):
        result = [model for model in result if model.get("vendorCode") == vendor_code]
    if region_code := _filter_value(standard_filters, "regionCode", "region_code"):
        result = [model for model in result if model.get("regionCode") == region_code]
    if family_code := _filter_value(standard_filters, "familyCode", "family_code"):
        result = [model for model in result if model.get("familyCode") == family_code]
    if capability := _filter_value(standard_filters, "capability"):
        result = [model for model in result if capability in model.get("capabilities", [])]
    if input_modality := _filter_value(standard_filters, "inputModality", "input_modality"):
        result = [model for model in result if input_modality in model.get("inputModalities", [])]
    if output_modality := _filter_value(standard_filters, "outputModality", "output_modality"):
        result = [model for model in result if output_modality in model.get("outputModalities", [])]
    if release_stage := _filter_value(standard_filters, "releaseStage", "release_stage"):
        result = [model for model in result if model.get("releaseStage") == release_stage]
    if shelf_state := _filter_value(standard_filters, "shelfState", "shelf_state"):
        result = [model for model in result if model.get("shelfState") == shelf_state]
    if routing_state := _filter_value(standard_filters, "routingState", "routing_state"):
        result = [model for model in result if model.get("routingState") == routing_state]
    if api_format := _filter_value(standard_filters, "apiFormat", "api_format"):
        result = [model for model in result if model.get("apiFormat") == api_format]
    return result


def list_available_models(catalog: ModelCatalog, filter: dict | None = None, **filters: str) -> list[dict]:
    standard_filters = dict(filter or {})
    standard_filters.update(filters)
    standard_filters["routingState"] = "enabled"
    standard_filters["shelfState"] = "listed"
    return [
        model
        for model in list_models(catalog, standard_filters)
        if get_model_prices(catalog, model.get("catalogKey", ""))
    ]


def _filter_value(filters: dict, *keys: str) -> str | None:
    for key in keys:
        value = filters.get(key)
        if isinstance(value, str) and value:
            return value
    return None


def catalog_key(vendor_code: str, region_code: str, model_id: str) -> str:
    return f"{vendor_code}/{region_code}/{model_id}"


def list_meters(catalog: ModelCatalog) -> list[dict]:
    return catalog.meters


def find_meter(catalog: ModelCatalog, meter_code: str) -> dict | None:
    return next((meter for meter in catalog.meters if meter.get("meterCode") == meter_code), None)


def find_model(catalog: ModelCatalog, catalog_key_value: str) -> dict | None:
    parts = catalog_key_value.split("/")
    if len(parts) != 3 or not parts[0] or not parts[1] or not parts[2]:
        return None
    return find_model_by_vendor_region(catalog, parts[0], parts[1], parts[2])


def find_model_by_vendor_region(catalog: ModelCatalog, vendor_code: str, region_code: str, model_id: str) -> dict | None:
    return next(
        (
            model
            for model in catalog.models
            if model.get("vendorCode") == vendor_code
            and model.get("regionCode") == region_code
            and model.get("modelId") == model_id
        ),
        None,
    )


def get_model_prices(catalog: ModelCatalog, catalog_key_value: str) -> list[dict]:
    parts = catalog_key_value.split("/")
    if len(parts) != 3 or not parts[0] or not parts[1] or not parts[2]:
        return []
    vendor_code, region_code, model_id = parts
    pricing = next(
        (
            item
            for item in catalog.pricing
            if item.get("vendorCode") == vendor_code
            and item.get("regionCode") == region_code
            and item.get("modelId") == model_id
        ),
        None,
    )
    return [] if pricing is None else pricing.get("prices", [])


def get_best_reference_price(catalog: ModelCatalog, catalog_key_value: str, meter_code: str) -> dict | None:
    return next((price for price in get_model_prices(catalog, catalog_key_value) if price.get("meterCode") == meter_code), None)


def list_models_by_capability(catalog: ModelCatalog, capability: str) -> list[dict]:
    return list_models(catalog, capability=capability)


def list_models_by_modality(catalog: ModelCatalog, input_modality: str, output_modality: str) -> list[dict]:
    return list_models(catalog, input_modality=input_modality, output_modality=output_modality)

File: sdkwork-models-flutter/lib/src/query.dart
import 'types.dart';

List<JsonObject> listVendors(ModelCatalog catalog) {
  final vendors = <String, JsonObject>{};
  for (final vendor in catalog.vendors) {
    final vendorCode = vendor['vendorCode'];
    if (vendorCode is! String || vendors.containsKey(vendorCode)) {
      continue;
    }
    vendors[vendorCode] = {
      'vendorCode': vendorCode,
      'displayName': vendor['displayName'],
      'legalName': vendor['legalName'],
      'vendorType': vendor['vendorType'],
      'capabilities': ((vendor['capabilities'] as List?) ?? const []).cast<Object?>(),
      'supportedProtocols': ((vendor['supportedProtocols'] as List?) ?? const []).cast<Object?>(),
      'openSource': vendor['openSource'] ?? false,
    };
  }
  return vendors.values.toList();
}

List<JsonObject> listVendorRegions(ModelCatalog catalog) {
  final regions = <JsonObject>[];
  for (final vendor in catalog.vendors) {
    final vendorCode = vendor['vendorCode'];
    if (vendorCode is! String) {
      continue;
    }
    final vendorRegions = vendor['regions'];
    if (vendorRegions is List) {
      for (final region in vendorRegions) {
        if (region is JsonObject && region['regionCode'] is String) {
          regions.add({'vendorCode': vendorCode, 'regionCode': region['regionCode']});
        }
      }
      continue;
    }
    final regionCode = vendor['regionCode'];
    if (regionCode is String) {
      regions.add({'vendorCode': vendorCode, 'regionCode': regionCode});
    }
  }
  return regions;
}

List<JsonObject> listModels(
  ModelCatalog catalog, {
  JsonObject? filter,
  String? vendorCode,
  String? regionCode,
  String? familyCode,
  String? capability,
  String? inputModality,
  String? outputModality,
  String? releaseStage,
  String? shelfState,
  String? routingState,
  String? apiFormat,
}) {
  return listModelsWhere(
    catalog,
    vendorCode: vendorCode ?? _filterString(filter, 'vendorCode'),
    regionCode: regionCode ?? _filterString(filter, 'regionCode'),
    familyCode: familyCode ?? _filterString(filter, 'familyCode'),
    capability: capability ?? _filterString(filter, 'capability'),
    inputModality: inputModality ?? _filterString(filter, 'inputModality'),
    outputModality: outputModality ?? _filterString(filter, 'outputModality'),
    releaseStage: releaseStage ?? _filterString(filter, 'releaseStage'),
    shelfState: shelfState ?? _filterString(filter, 'shelfState'),
    routingState: routingState ?? _filterString(filter, 'routingState'),
    apiFormat: apiFormat ?? _filterString(filter, 'apiFormat'),
  );
}

List<JsonObject> listModelsWhere(
  ModelCatalog catalog, {
  String? vendorCode,
  String? regionCode,
  String? familyCode,
  String? capability,
  String? inputModality,
  String? outputModality,
  String? releaseStage,
  String? shelfState,
  String? routingState,
  String? apiFormat,
}) {
  return catalog.models.where((model) {
    if (vendorCode != null && model['vendorCode'] != vendorCode) {
      return false;
    }
    if (regionCode != null && model['regionCode'] != regionCode) {
      return false;
    }
    if (familyCode != null && model['familyCode'] != familyCode) {
      return false;
    }
    if (capability != null) {
      final capabilities = (model['capabilities'] as List?) ?? const [];
      if (!capabilities.contains(capability)) {
        return false;
      }
    }
    if (inputModality != null) {
      final inputModalities = (model['inputModalities'] as List?) ?? const [];
      if (!inputModalities.contains(inputModality)) {
        return false;
      }
    }
    if (outputModality != null) {
      final outputModalities = (model['outputModalities'] as List?) ?? const [];
      if (!outputModalities.contains(outputModality)) {
        return false;
      }
    }
    if (releaseStage != null && model['releaseStage'] != releaseStage) {
      return false;
    }
    if (shelfState != null && model['shelfState'] != shelfState) {
      return false;
    }
    if (routingState != null && model['routingState'] != routingState) {
      return false;
    }
    if (apiFormat != null && model['apiFormat'] != apiFormat) {
      return false;
    }
    return true;
  }).toList();
}

List<JsonObject> listAvailableModels(
  ModelCatalog catalog, {
  JsonObject? filter,
  String? vendorCode,
  String? regionCode,
  String? familyCode,
  String? capability,
  String? inputModality,
  String? outputModality,
  String? releaseStage,
  String? apiFormat,
}) {
  return listModels(
    catalog,
    filter: filter,
    vendorCode: vendorCode,
    regionCode: regionCode,
    familyCode: familyCode,
    capability: capability,
    inputModality: inputModality,
    outputModality: outputModality,
    releaseStage: releaseStage,
    shelfState: 'listed',
    routingState: 'enabled',
    apiFormat: apiFormat,
  ).where((model) {
    final catalogKeyValue = model['catalogKey'];
    return catalogKeyValue is String && getModelPrices(catalog, catalogKeyValue).isNotEmpty;
  }).toList();
}

String catalogKey(String vendorCode, String regionCode, String modelId) => '$vendorCode/$regionCode/$modelId';

List<JsonObject> listMeters(ModelCatalog catalog) => catalog.meters;

JsonObject? findMeter(ModelCatalog catalog, String meterCode) {
  for (final meter in catalog.meters) {
    if (meter['meterCode'] == meterCode) {
      return meter;
    }
  }
  return null;
}

JsonObject? findModel(ModelCatalog catalog, String catalogKeyValue) {
  final parts = catalogKeyValue.split('/');
  if (parts.length != 3 || parts[0].isEmpty || parts[1].isEmpty || parts[2].isEmpty) {
    return null;
  }
  return findModelByVendorRegion(catalog, parts[0], parts[1], parts[2]);
}

JsonObject? findModelByVendorRegion(ModelCatalog catalog, String vendorCode, String regionCode, String modelId) {
  for (final model in catalog.models) {
    if (model['vendorCode'] == vendorCode && model['regionCode'] == regionCode && model['modelId'] == modelId) {
      return model;
    }
  }
  return null;
}

List<JsonObject> getModelPrices(ModelCatalog catalog, String catalogKeyValue) {
  final parts = catalogKeyValue.split('/');
  if (parts.length != 3 || parts[0].isEmpty || parts[1].isEmpty || parts[2].isEmpty) {
    return const [];
  }
  final vendorCode = parts[0];
  final regionCode = parts[1];
  final modelId = parts[2];
  for (final item in catalog.pricing) {
    if (item['vendorCode'] == vendorCode && item['regionCode'] == regionCode && item['modelId'] == modelId) {
      return ((item['prices'] as List?) ?? const []).cast<JsonObject>();
    }
  }
  return const [];
}

JsonObject? getBestReferencePrice(ModelCatalog catalog, String catalogKeyValue, String meterCode) {
  for (final price in getModelPrices(catalog, catalogKeyValue)) {
    if (price['meterCode'] == meterCode) {
      return price;
    }
  }
  return null;
}

List<JsonObject> listModelsByCapability(ModelCatalog catalog, String capability) {
  return listModelsWhere(catalog, capability: capability);
}

List<JsonObject> listModelsByModality(ModelCatalog catalog, String inputModality, String outputModality) {
  return listModelsWhere(catalog, inputModality: inputModality, outputModality: outputModality);
}

List<JsonObject> listProtocols(ModelCatalog catalog) {
  return catalog.protocols;
}

JsonObject? findProtocol(ModelCatalog catalog, String protocolCode) {
  for (final p in catalog.protocols) {
    if (p['protocolCode'] == protocolCode) {
      return p;
    }
  }
  return null;
}

List<JsonObject> listProtocolsByVendor(ModelCatalog catalog, String vendorCode) {
  final vendor = catalog.vendors.cast<JsonObject?>().firstWhere(
        (v) => v?['vendorCode'] == vendorCode,
        orElse: () => null,
      );
  if (vendor == null) {
    return const [];
  }
  final supported = ((vendor['supportedProtocols'] as List?) ?? const [])
      .whereType<String>()
      .toSet();
  return catalog.protocols.where((p) => supported.contains(p['protocolCode'])).toList();
}

List<JsonObject> listModelsByProtocol(ModelCatalog catalog, String protocolCode) {
  return listModelsWhere(catalog, apiFormat: protocolCode);
}

String? _filterString(JsonObject? filter, String key) {
  final value = filter?[key];
  return value is String && value.isNotEmpty ? value : null;
}
